### ISSUE TITLE
we only support the cyruslibs version of cld2

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#3131]
The cyruslibs version of cld2 has an extra commit on top which provides a bunch of necessary installation stuff that I guess the upstream repository doesn't provide:  https://github.com/cyrusimap/cld2/commit/e8318941ac320ab8c019dfa52b0401ec3d3bccb2

But, distros have handle this differently (and possibly mutually incompatibly, I'm not sure).

For now, the Debian and Fedora(?) libcld2 packages are known to not be compatible with Cyrus, because they reproduce the "public" vs "internal" header subdivision that the upstream repository uses, whereas the cyruslibs version puts them all in the same directory.  Cyrus depends on the cyruslibs organisation.  

It's also worth noting that the Debian and Fedora(?) packages also don't provide a .pc file (but the cyruslibs one does).

For now, we simply must use the cyruslibs version.  This is fine for the unstable branch, but we'll need to consolidate this in some coherent way for the 3.4 release.

**Comment by elliefm** [comment]
@ksmurchison  I think I remember that you run Fedora, but please contradict me (or just edit the OP) if I'm wrong.  Thanks :)

**Comment by ksmurchison** [comment]
Current Fedora 32 dev package:

Name        : cld2-devel
Version     : 0
Release     : 0.15.20150821gitb56fa78.fc32
Architecture: x86_64
Install Date: Sat 02 May 2020 09:27:22 AM EDT
Group       : Unspecified
Size        : 1346857
License     : ASL 2.0
Signature   : RSA/SHA256, Tue 28 Jan 2020 09:28:10 AM EST, Key ID 6c13026d12c944d0
Source RPM  : cld2-0-0.15.20150821gitb56fa78.fc32.src.rpm
Build Date  : Tue 28 Jan 2020 09:13:37 AM EST
Build Host  : buildvm-27.phx2.fedoraproject.org
Packager    : Fedora Project
Vendor      : Fedora Project
URL         : https://github.com/CLD2Owners/cld2/
Bug URL     : https://bugz.fedoraproject.org/cld2
Summary     : Development files for cld2
Description :
A library that detects over 80 languages in UTF-8 text, based largely
on groups of four letters. Also tables for 160+ language versions.

This sub-package contains the headers for cld2.
/usr/include/cld2
/usr/include/cld2/internal
/usr/include/cld2/internal/cld2_dynamic_compat.h
/usr/include/cld2/internal/cld2_dynamic_data.h
/usr/include/cld2/internal/cld2_dynamic_data_extractor.h
/usr/include/cld2/internal/cld2_dynamic_data_loader.h
/usr/include/cld2/internal/cld2tablesummary.h
/usr/include/cld2/internal/cldutil.h
/usr/include/cld2/internal/cldutil_offline.h
/usr/include/cld2/internal/cldutil_shared.h
/usr/include/cld2/internal/compact_lang_det_hint_code.h
/usr/include/cld2/internal/compact_lang_det_impl.h
/usr/include/cld2/internal/debug.h
/usr/include/cld2/internal/fixunicodevalue.h
/usr/include/cld2/internal/generated_language.h
/usr/include/cld2/internal/generated_ulscript.h
/usr/include/cld2/internal/getonescriptspan.h
/usr/include/cld2/internal/integral_types.h
/usr/include/cld2/internal/lang_script.h
/usr/include/cld2/internal/langspan.h
/usr/include/cld2/internal/offsetmap.h
/usr/include/cld2/internal/port.h
/usr/include/cld2/internal/scoreonescriptspan.h
/usr/include/cld2/internal/stringpiece.h
/usr/include/cld2/internal/tote.h
/usr/include/cld2/internal/unittest_data.h
/usr/include/cld2/internal/utf8acceptinterchange.h
/usr/include/cld2/internal/utf8prop_lettermarkscriptnum.h
/usr/include/cld2/internal/utf8repl_lettermarklower.h
/usr/include/cld2/internal/utf8scannot_lettermarkspecial.h
/usr/include/cld2/internal/utf8statetable.h
/usr/include/cld2/public
/usr/include/cld2/public/compact_lang_det.h
/usr/include/cld2/public/encodings.h
/usr/lib64/libcld2.so
/usr/lib64/libcld2_dynamic.so
/usr/lib64/libcld2_full.so
/usr/share/doc/cld2-devel
/usr/share/doc/cld2-devel/CLD2UnitTestFullOutput.html
/usr/share/doc/cld2-devel/CLD2UnitTestOutput.html
/usr/share/doc/cld2-devel/CLD2UnitTestOutputVerbose.html
/usr/share/doc/cld2-devel/a_little_french_test_input.html
/usr/share/doc/cld2-devel/evaluate_cld1_small_20110406.txt
/usr/share/doc/cld2-devel/evaluate_cld2_large_20130720.txt
/usr/share/doc/cld2-devel/evaluate_cld2_large_20140122.txt
/usr/share/doc/cld2-devel/evaluate_cld2_small_20130715.txt
/usr/share/doc/cld2-devel/evaluate_cld2_small_20140122.txt
/usr/share/doc/cld2-devel/test_version.html
/usr/share/doc/cld2-devel/test_version.txt

**Comment by rsto** [comment]
The way we organize the headers in our CLD2 cyruslibs fork was done when I experimented with the library to see if we can use it in Cyrus. Now that language indexing will get into the next stable release, I agree that we should make it work with packaged cld2, too.I will see see to update the installer and Cyrus autoconf for that.

**Comment by elliefm** [comment]
Our configure.ac looks fine, in that it just uses pkg-config to find it... if only the distro packages would be kind enough to install a pc file like the cyruslibs one does.

If they don't, and we need to provide an oldschool handwritten --with-cld2=path configuration, that would be a bit of a nuisance.  I wonder how hard it would be to get the distros to patch in a pc file?

If we can get the cyruslibs install layout compatible with theirs, and get our pc file to work with that, then they could simply copy our pc file work.  It's not like it's a hard technical problem to solve, it's more of a "can we deliver a patch to them and will they accept it".  I don't see why not, but I don't have any experience contributing to distro packages, so idk.

**Comment by ksmurchison** [comment]
I believe rsto fixed this on master

**Comment by rsto** [comment]
Yes, I fixed that in https://github.com/cyrusimap/cld2/commit/074b451c85dc6a97930273147695abb735a1626a. But we still require pkg-config to find CLD2.

**Comment by rjbs** [comment]
We no longer require a special version cld2, see cyruslibs@6575d54eefe317bb8919c31c5d5b4e6f501fab9c
